### ISSUE TITLE
let_and_return raised on conditional shadowing

### ISSUE THREAD

**Issue by benschulz** [rust-lang/rust-clippy#10364]
### Summary

On a let that *conditionally* shadows the previous, let_and_return raises a false positive.

```rust
let result = foo;
#[cfg(some_cfg)]
let result = bar(result);
result
```

This can be worked around by using the less DRY version from #1346.
```rust
#[cfg(not(some_cfg))]
let result = foo;
#[cfg(some_cfg)]
let result = bar(foo);
result
```

### Lint Name

let_and_return

### Reproducer

I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0eae21d8d2f7c00688787d0b5fedd4a8)):

```rust
pub fn one_or_two() -> u32 {
    let result = 1;

    #[cfg(feature = "two")]
    let result = 2 * result;

    result
}
```

I saw this happen:

```
warning: returning the result of a `let` binding from a block
 --> src/lib.rs:8:5
  |
3 |     let result = 1;
  |     --------------- unnecessary `let` binding
...
8 |     result
  |     ^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_and_return
  = note: `#[warn(clippy::let_and_return)]` on by default
help: return the expression directly
  |
3 ~     
4 | 
...
7 | 
8 ~     1
  |
```

I expected to see this happen:
No lint emitted.

### Version

```text
rustc 1.69.0-nightly (5b8f28453 2023-02-12)
binary: rustc
commit-hash: 5b8f284536d00ba649ca968584bedab4820d8527
commit-date: 2023-02-12
host: x86_64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```

**Comment by jplatte** [comment]
Duplicate of #9150.
